### PR TITLE
Allow modifications to VM spec during VM snapshot, except disks/volumes

### PR DESCRIPTION
**Pull request by awels**
### What this PR does
Before this PR:
Any modification attempt to the VM spec is rejected while a VM snapshot status is set. This can cause the finalizer to not be allowed to be set after the status is updated. For instance with an older VM spec that doesn't contain the architecture field. There is a mutating webhook that automatically adds the architecture field to the spec. However the status indicates that a snapshot is in progress and it will reject changes to the VM spec.

Now the snapshot can never complete because it can't get a 'lock' on the VM (which requires the finalizer).

After this PR:
Relaxed the requirement that any changes to the VM spec are rejected. Only changes to the disks/volumes are rejected. This prevents the above scenario from happening. 

Fixes https://issues.redhat.com/browse/CNV-48769

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: 

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [x] PR: The PR description is expressive enough and will help future contributors
- [x] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [x] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
Relaxed check on modify VM spec during VM snapshot to only check disks/volumes
```

### CODE REVIEW

File: pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go [hunk 1]
```diff
@@ -779,13 +779,42 @@ func validateSnapshotStatus(ar *admissionv1.AdmissionRequest, vm *v1.VirtualMach
...
-			Message: fmt.Sprintf("Cannot update VM spec until snapshot %q completes", *vm.Status.SnapshotInProgress),
+			Message: fmt.Sprintf("Cannot update VM disks or volumes until snapshot %q completes", *vm.Status.SnapshotInProgress),
 			Field:   k8sfield.NewPath("spec").String(),
 		}}
 	}
 
 	return nil
 }
+
+func compareDisks(old, new []v1.Disk) bool {
```
Comment (mhenriks) [inline pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go:794]: Not sure we care about disks only volumes: https://github.com/kubevirt/kubevirt/blob/b62599eee3960fb389330c870d1ca71a987e6a21/pkg/storage/snapshot/source.go#L359

**Comment by ShellyKa13** [comment]
The finalizer shouldnt have a problem to be added since it is not part of the spec..
I understand that when trying to put the finalizer a mutating webhook tries to also change the spec and so this update is rejected. With @mhenriks PR he changes this update to be Patch: https://github.com/kubevirt/kubevirt/pull/12601/files#diff-98499752dcfec117369933a187ef821e70589d979cba2b3b5cdfa1d9149be198R127
Will this prevent from the mutating webhook to be applied?
Not sure why would we want to allow changes in vm spec during snapshot, since we also save the spec in the process user will not know if his changes will be included in the snapshot or not..

**Comment by awels** [comment]
If the patch PR fixes the problem, then we can close this PR. Just have to verify that the mutating webhook is not applied to the patch, which if I read the documetation correctly is the case.

**Comment by mhenriks** [comment]
> If the patch PR fixes the problem, then we can close this PR. Just have to verify that the mutating webhook is not applied to the patch, which if I read the documetation correctly is the case.

The webhook will still get called on `Patch`

> Not sure why would we want to allow changes in vm spec during snapshot, since we also save the spec in the process user will not know if his changes will be included in the snapshot or not..

Yeah allowing spec update could be confusing to users but I don't think it's a huge deal. 

The spec that is stored as part of the snapshot is saved at the beginning of the operation when VMSnapshotContent is created. I am wondering if the VM spec changing after that could somehow break the rest of the process. It just occurred to me that changing `running`/`runStrategy` could be problematic? But maybe not? Anything else @ShellyKa13?

**Comment by mhenriks** [comment]
> > The webhook will still get called on Patch
> 
> Thats too bad.. I think that we dont want to let running and runstrategy change during the snapshot since change like that can cause a vm to start during snapshot while the snapshot is not marked as online.. And regarding the rest of the spec Im not sure either.. cant tell what else can cause an issue, maybe nothing, but dont know.. I believe its safer to not let the spec change but with the bug not sure whats the other option here..

We *could* enumerate all the fields that may be updated in a mutating webhook and ignore them but keeping that list up to date would be really annoying.

I just scanned the `vmSnapshotSource` and I think we really only care about volumes (directly) and running/runStrategy (indirectly)

**Pull request merged**